pkg/controller/metrics/metricsWithDynamicLabels.go

```diff
@@ -0,0 +1,91 @@
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+)
+
+// metricsWithDyanamicLabels encompasses metrics that can have optional labels.
+type metricsWithDyanamicLabels interface {
+	LogMetricWithDynamicLabels(metav1.Object, log.FieldLogger, float64)
+}
+
+type CounterWithDynamicLabels struct {
+	*prometheus.CounterVec
+	FixedLabels    map[string]string
+	OptionalLabels map[string]string
+}
+
+// LogMetricWithDynamicLabels logs the counter metric. It simply increments the value, so last parameter isn't used.
+func (c CounterWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, _ float64) {
+	curriedVec, err := c.CounterVec.CurryWith(c.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(c.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, c.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Inc()
+	}
+}
+
+type HistogramWithDynamicLabels struct {
+	*prometheus.HistogramVec
+	FixedLabels    map[string]string
+	OptionalLabels map[string]string
+}
+
+// LogMetricWithDynamicLabels observes the histogram metric with val value.
+func (h HistogramWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, val float64) {
+	curriedVec, err := h.HistogramVec.CurryWith(h.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(h.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, h.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Observe(val)
+	}
+}
+
+var _ metricsWithDyanamicLabels = CounterWithDynamicLabels{}
+var _ metricsWithDyanamicLabels = HistogramWithDynamicLabels{}
+
+// GetOptionalClusterTypeLabels reads the AdditionalClusterDeploymentLabels from the metrics config and returns the same
+// as a map
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) map[string]string {
+	var mapClusterTypeLabelToValue = make(map[string]string)
```

Comment:
nit: You're not really gaining anything by `make()`ing this map, unless you specify a size.